source/modulo_interfaces/msg/Predicate.msg

```diff
@@ -0,0 +1,5 @@
+string component
+string controller
+string hardware
```

Comment:
For consumers of the predicate message, I was originally considering checking the contents like so:

```
if (msg->component)
  // handle the predicate from the component
else if (msg->controller)
  // handle the predicate based on msg->controller and msg->hardware
```

But, maybe using separate string fields and leaving them either empty or filled is not the nicest design. It's easy to break the logic if you fill all fields, wrong combination of fields, or no fields.

The alternative (just sending the fully-qualified node name of the producer) was simpler, but made more work for the consumer of the predicate to determine the nature of the producer. I think we can have a nice middle ground if we include a type field to specify the nature of the producer, and then send the fully qualified node name with it.

```suggestion
uint8 COMPONENT=1
uint8 CONTROLLER=2

uint8 type
string node
```

This pattern with exclusive types is also a bit more scalable (e.g. if somehow we add hardware predicates or other sources of predicates)

```
switch(msg->type) {
  case Predicate::COMPONENT: 
    // handle the predicate from the component
  case Predicate::CONTROLLER:
    // handle the predicate based on msg.controller and msg.hardware
}
```